eclipse-theia/theia
# Support language overrides in frontend config object

#### What it does


Fixes #8314 by reading the the frontend config's `preferences` object as though it were a `settings.json` object.

#### How to test


1. Add a language-specific setting to the frontend object in the `package.json` of the browser or electron example apps. E.g.
```
  "theia": {
    "target": "electron",
    "frontend": {
      "config": {
        "applicationName": "Theia Electron Example"
        "preferences": {
          "[typescript]": {
            "editor.fontSize": 20
          }
        }
      }
    }
  },
```
2. Start appropriate version of the application.
3. Make sure you don't have any other settings conflicting with the default set in the `package.json` (e.g. a TS-specific font size in your user scope)
3. Open a file of the language for which you set a different default.
4. Observe that the language-specific default setting takes effect.
5. Open a file in a different language.
6. Observe that the language-specific default does not apply.

#### Caveats
 - In order to move the `getParsedContent` up the inheritance tree into the `PreferenceProvider` abstract class, I changed its API to accept an object rather than a string.
 - I also needed to move the `testOverrideValue` method outside the `PreferenceSchemaProvider` class, and so I extracted various override-related functionality into a separate service (`PreferenceLanguageOverrideService`). If that seems like too large a change in the behavior of the `PreferenceProvider`, much of that functionality can be moved back, as long as `testOverrideValue` can be used by the base `PreferenceProvider` class.
 - If a `PreferenceContribution` or plugin adds a language-specific override (e.g. YAML's two-space tabs size), it will override existing overrides in the `package.json`. If we prefer to give priority to the `package.json`, the logic can be rearranged to respect the settings found there.

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: Colin Grant <colin.grant@ericsson.com>
